docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,354 @@
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+A new component named gardener-custom-metrics is deployed on each seed. It directly scrapes metrics data from all
+ShootKapi pods on the seed and derives custom metrics based on it. The component is registered as an extension API
+service to the seed kube-apiserver. It occupies the custom metrics API extension point and is responsible for providing
+all custom metrics for the seed kube-apiserver, including the one driving the ShootKapi HPA.
+
+![Design outline](./assets/gep22-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: Gardener Custom Metrics Provider Component 
+A new component, named gardener-custom-metrics is added to seed clusters. It periodically scrapes the metrics endpoints
+of all ShootKapi pods on the seed. It implements the K8s custom metrics API and provides K8s metrics specific to
+Gardener, based on custom calculations. The proposed design can naturally be extended to multiple sources of
+input data and a non-volatile cache for acquired data. However, in the scope of this proposal, the only data source
+is the aforementioned metrics scrape, and the calculated values only need to be briefly cached in memory.
+
+A single replica is deployed, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
+
+A kube-apiserver's metrics response measures megabytes in size. To reduce the amount of network traffic, this proposal
+utilizes compressed HTTP responses when scraping metrics. Outside the scope of this proposal, a future enhancement is
+possible, in which a metrics filter component is added as a sidecar to each ShootKapi pod. Each such sidecar would
+mirror the metrics endpoint of its respective ShootKapi, with the only addition of support for filtering based on an
+HTTP request parameter (e.g. `GET /metrics?name=apiserver_request_total` would forward a `GET /metrics` request to the
+ShootKapi, and then would reduce the response to only `apiserver_request_total` counters, before passing it on to the
+caller). 
+
+### Element: New Custom Pod Metric for ShootKapis
+A new `shoot:apiserver_request_total:rate_by_pod` pod custom metric is made available for each ShootKapi pod on the
```

Comment:
I would call the custom metric something like `apiserver_request_total:rate`. It isn't limited to shoots as such and `by_pod` is implied by the nature of a pod custom metric.
But at the end, this is an implementation detail and it doesn't have to be agreed upon right now.

Reply:
Good point! After the transition to gardener-custom-metrics, the metric no longer lives in Prometheus, so it's just a custom pod metric and  the "_pod" suffix is pointless.

Fixed and will be available in next push.